kubeadm - fix upgrades with external etcd

detiber opened the pull request:
**What this PR does / why we need it**:

- Allow for upgrade plan and upgrade apply to work with external etcd
  - https://github.com/kubernetes/kubeadm/issues/727
  - https://github.com/kubernetes/kubernetes/pull/62141

- Update upgrade plan output when configured for external etcd
  - Move etcd to a separate section and show available upgrades

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes/kubeadm/issues/727

**Release note**:
```release-note
kubeadm upgrade now supports external etcd setups again
```

I created a gist documenting the manual testing I've been doing for this PR here: https://gist.github.com/detiber/e18d907c41901fbb5e12ffa1af5750f8

craigtracey committed:
Support kubeadm upgrade with remote etcd cluster

Currently kubeadm only performs an upgrade if the etcd cluster is
colocated with the control plane node. As this is only one possible
configuration, kubeadm should support upgrades with etcd clusters
that are not local to the node.

Signed-off-by: Craig Tracey <craigtracey@gmail.com>

detiber commented:
/cc @kubernetes/sig-cluster-lifecycle-pr-reviews

detiber commented:
/assign @craigtracey

detiber commented:
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-e2e-gce-device-plugin-gpu

timothysc commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -51,6 +51,7 @@ type applyFlags struct {
 	force              bool
 	dryRun             bool
 	etcdUpgrade        bool
+	externalEtcd       bool
```
I don't see why we need a var.  We only support static manifest so triggering on that seems sufficient.

detiber replied:
Accidental cruft left from rebasing and cleaning this up after sitting for a while, will remove.

timothysc commented on cmd/kubeadm/app/cmd/upgrade/plan.go:
```diff
@@ -89,14 +89,34 @@ func RunPlan(flags *planFlags) error {
+	isExternalEtcd := len(upgradeVars.cfg.Etcd.Endpoints) > 0
+	if isExternalEtcd {
+		// Create external etcd client
+		client, err := etcdutil.NewClient(
```
can't we do this for * in 1.11?

detiber replied:
Not quite, since we need to introspect the static pod manifest for the certs, which is what NewStaticPodClient does.

timothysc reviewed (changes requested):
Also should baseline from https://github.com/kubernetes/kubernetes/pull/62141 for attribution and address those comments as well.

detiber committed:
kubeadm - fix external etcd upgrades

- Update upgrade plan output when configured for external etcd
  - Move etcd to a separate section and show available upgrades

detiber commented:
Updated to rebuild on top of https://github.com/kubernetes/kubernetes/pull/62141

detiber commented:
@craigtracey ptal

timothysc reviewed (approved):
/lgtm 
/approve 

There are nits, but let's not hold it up.  Major behavior is there.  Thx @detiber!

detiber commented:
rebased to resolve merge conflict

luxas reviewed (approved):
/lgtm

stealthybox commented:
@detiber thanks for taking the time to get this in -- this is a really clean refactor :+1:
